### ISSUE TITLE
Specifying reason in `expect(clippy::needless_return)` triggers false positive "this lint expectation is unfulfilled"

### ISSUE THREAD

**Issue by schuetzm** [rust-lang/rust-clippy#13366]
### Summary

It seems the `reason` option has an effect on whether the lint warns about an unfulfilled expectation.

### Lint Name

clippy::needless_return

### Reproducer

I tried this code:

```rust
fn a(x: Option<u8>) -> Option<u8> {
    match x {
        Some(_) => None,
        None => {
            #[expect(clippy::needless_return, reason = "Use early return for errors.")]
            return None;
        }
    }
}

fn b(x: Option<u8>) -> Option<u8> {
    match x {
        Some(_) => None,
        None => {
            #[expect(clippy::needless_return)]
            return None;
        }
    }
}
```

I saw this happen:
The warning about unfulfilled expectations is printed for function `a`, but not for `b`:

```
  |
5 |             #[expect(clippy::needless_return, reason = "Use early return for errors.")]
  |                      ^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: Use early return for errors.
  = note: `#[warn(unfulfilled_lint_expectations)]` on by default
```

I expected to see this happen:
No warning should be printed for either function.

### Version

```text
rustc 1.81.0 (eeb90cda1 2024-09-04)
binary: rustc
commit-hash: eeb90cda1969383f56a2637cbd3037bdf598841c
commit-date: 2024-09-04
host: x86_64-unknown-linux-gnu
release: 1.81.0
LLVM version: 18.1.7
```

**Comment by Jarcho** [comment]
This lint has a hack to make`#[expect(..)]` work since normally it doesn't trigger when an attribute is present on the statement. It needs to be extended to ignore the reason field.
